internal/provider/resource_tfe_workspace_settings.go

```diff
@@ -0,0 +1,410 @@
+					stringvalidator.RegexMatches(
+						IDPattern("apool"),
+						"must be a valid workspace ID (apool-<RANDOM STRING>)",
+					),
+				},
+			},
+
+			// ListAttribute was required here because we are still using plugin protocol v5.
+			// Once compatibility is broken for v1, and we convert all
+			// providers to protocol v6, this can become a single nested object.
+			"overwrites": schema.ListAttribute{
+				Computed:    true,
+				ElementType: overwritesElementType,
+				PlanModifiers: []planmodifier.List{
+					revertOverwritesIfExecutionModeUnset{},
+				},
+			},
+		},
+	}
+}
+
+// workspaceSettingsModelFromTFEWorkspace builds a resource model from the TFE model
+func workspaceSettingsModelFromTFEWorkspace(ws *tfe.Workspace) *modelWorkspaceSettings {
+	result := modelWorkspaceSettings{
+		ID:            types.StringValue(ws.ID),
+		WorkspaceID:   types.StringValue(ws.ID),
+		ExecutionMode: types.StringValue(ws.ExecutionMode),
+	}
+
+	if ws.AgentPool != nil && ws.ExecutionMode == "agent" {
+		result.AgentPoolID = types.StringValue(ws.AgentPool.ID)
+	}
+
+	settingsModel := modelOverwrites{
+		ExecutionMode: types.BoolValue(false),
+		AgentPool:     types.BoolValue(false),
+	}
+
+	if ws.SettingOverwrites != nil {
+		settingsModel = modelOverwrites{
+			ExecutionMode: types.BoolValue(*ws.SettingOverwrites.ExecutionMode),
+			AgentPool:     types.BoolValue(*ws.SettingOverwrites.AgentPool),
+		}
+	}
+
+	listOverwrites, diags := types.ListValueFrom(ctx, overwritesElementType, []modelOverwrites{settingsModel})
+	if diags.HasError() {
+		panic("Could not build list value from slice of models. This should not be possible unless the model breaks reflection rules.")
+	}
+
+	result.Overwrites = listOverwrites
+
+	return &result
+}
+
+func (r *workspaceSettings) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data modelWorkspaceSettings
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	model, err := r.readSettings(ctx, data.WorkspaceID.ValueString())
+	if errors.Is(err, errWorkspaceNoLongerExists) {
+		resp.State.RemoveResource(ctx)
+		return
+	} else if err != nil {
+		resp.Diagnostics.AddError("Error reading workspace", err.Error())
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, model)...)
+}
+
+func (r *workspaceSettings) readSettings(ctx context.Context, workspaceID string) (*modelWorkspaceSettings, error) {
+	ws, err := r.config.Client.Workspaces.ReadByID(ctx, workspaceID)
+	if err != nil {
+		// If it's gone: that's not an error, but we are done.
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Workspace %s no longer exists", workspaceID)
+			return nil, errWorkspaceNoLongerExists
+		}
+		return nil, fmt.Errorf("couldn't read workspace %s: %s", workspaceID, err.Error())
+	}
+
+	return workspaceSettingsModelFromTFEWorkspace(ws), nil
+}
+
+func (r *workspaceSettings) updateSettings(ctx context.Context, data *modelWorkspaceSettings, state *tfsdk.State) error {
```

Comment:
Shouldn't we add handling here for outdated versions of TFE that do not support setting overwrites? I believe in this case, the provider would previously use `remote` as the default execution_mode, whereas I believe this code would simply NOT set the execution_mode.

A part of me really wants to force users to set the `execution_mode`... purely based on how that would drastically reduce the complexity of the provider from BOTH a user standpoint and a developer standpoint. That being said, it feels wrong to oppose the API and (effectively) disallow users the ability to use default execution modes if they use the terraform provider.

Reply:
Awesome 💯